_posts/2021-10-05-hhvm-4.130.markdown

```diff
@@ -0,0 +1,89 @@
+- `hh_client --lint` can now suggest variance; for example:
+
+```Hack
+    Lint[5639] This generic parameter could be marked covariant. Consider prefixing the generic parameter with + [1]
+
+    infer_variance.php:1:19
+    [1] 1 | interface IGetter<T> {
+        2 |   public function get(): T;
+        3 | }
+```
+
+
+
+# Breaking Changes
+
+- It is now an error to pass the first argument to `invariant()` as `inout`.
+- Using `self` and `parent` as types now results in a runtime error; this previously 
+  raised typechecker errors. They remain valid for use in other contexts, e.g.
+  `self::foo()` and `parent::__construct()`.
+- `array_map()` is no longer special-cased in the typechecker; it now expects
+  a single callback, a single `KeyedContainer`, and returns a `KeyedContainer`.
+  It no longer accepts a variable number of parameters, and no longer preserves
+  the concrete type of the input.
+- The `Str\` functions no longer respect the current thread locale; functions
+  that were locale-sensitive now have a `_l()` variant that takes an explicit
+  locale.
+  - For example, `Str\format('%0.2f', 1.23)` will now reliably return `1.23`,
+    instead of sometimes returning `1,23` - if the previous behavior is desired,
+    `Str\format_l(Locale\get_native(), '%0.2f', 1.23)` can be used instead.
+  - We made this change as implicit use of the locale meant that the majority
+    of `Str\` functions could not be marked as pure, and because most users
+    were surprised that functions like `Str\format()`, `sprintf()` etc were 
+    locale-aware.
+  - The `_l` functions support UTF-8, depending on the CTYPE in the locale; this
+    may lead to behavior differences. Please ensure that `LC_CTYPE` is set
+    correctly for the behavior you desire using `Locale\modified()`.
+- `Str\split()` now consistently throws if a negative limit was provided.
+  - previously, `Str\split()` would only throw for negative limits if the
+    delimiter was empty.
+  - the previous behavior was unintuitive, and inconsistent with other functions
+    that accept negative limits with offsets:
+    - `Str\split('a!b!c', '!')` returned `vec['a', 'b', 'c']`
+    - `Str\split('a!b!c', '!', 2)` returned `vec['a', 'b!c']`
+    - `Str\split('a!b!c', '!', -1)` instead returned `vec['a', 'b']`
+  - added `Legacy_FIXME\split_with_possibly_negative_limit()` as a migration aid
+    with the old behavior.
+- `Str\replace_every()` and related functions now consistently raises exceptions
+  for empty search strings, and for inputs that are not strings. Previous
+  behavior varied depending on the specific function and inputs.
+  - added `Legacy_FIXME\coerce_possibly_invalid_str_replace_pairs()` as a
+    migration aid.
+- `Str\` functions now raise `InvalidArgumentException` instead of invariant
+  violations for invalid arguments.
+
+# Future Changes
+
+- We may remove implicit coercion of enum values in a future release; if the
```

Comment:
Does this affect `EnumName::assert()`, `EnumName::isValid()`, `expr is EnumName`, and `expr as EnumName`? I am assuming this coercion is about intish string coercion being removed.

Reply:
The test case for this commit is unclear to me.
https://github.com/facebook/hhvm/commit/481b914c0e14ac0a430f65b8fcd46df06973312a

Reply:
Did some more testing and added explanation :)